estimator.go

```diff
@@ -0,0 +1,343 @@
+}
+
+func stubAddress() common.Address {
+	raw := make([]byte, 20)
+	rand.Read(raw)
+	return common.BytesToAddress(raw)
+}
+
+func (e *Estimator) BuildStubSignature(walletConfig WalletConfig, willSign []bool, isEoa []bool) []byte {
+	parts := make(SignatureParts, walletConfig.Signers.Len())
+
+	// pre-determined signature, tailored for worse-case scenario in gas costs
+	// TODO: Compute average siganture and present a more likely scenario for a more close estimation
+	sig := common.Hex2Bytes("1fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff7fffffffffffffffffffffffffffffff5d576e7357a4501ddfe92f46681b20a01b02")
+
+	for i, s := range walletConfig.Signers {
+		if willSign[i] {
+			if isEoa[i] {
+				parts[i] = &SignaturePart{
+					Weight:  s.Weight,
+					Address: s.Address,
+					Type:    sigPartTypeEOA,
+					Value:   sig,
+				}
+			} else {
+				sig := e.BuildStubSignature(WalletConfig{
+					Threshold: 2,
+					Signers: WalletConfigSigners{
+						WalletConfigSigner{
+							Address: stubAddress(),
+							Weight:  1,
+						},
+						WalletConfigSigner{
+							Address: stubAddress(),
+							Weight:  1,
+						},
+						WalletConfigSigner{
+							Address: stubAddress(),
+							Weight:  1,
+						},
+					},
+				}, []bool{false, true, true}, []bool{false, false, false})
+
+				sig = append(sig, 03)
+
+				parts[i] = &SignaturePart{
+					Weight:  s.Weight,
+					Address: s.Address,
+					Type:    sigPartTypeDynamic,
+					Value:   sig,
+				}
+			}
+		} else {
+			parts[i] = &SignaturePart{
+				Weight:  s.Weight,
+				Address: s.Address,
+				Type:    sigPartTypeAddress,
+			}
+		}
+	}
+
+	signature := Signature{
+		Threshold: walletConfig.Threshold,
+		Signers:   parts,
+	}
+
+	encoded, err := signature.Encode()
+
+	// This method builds a signature from scratch
+	// it should never faild to encode it
+	if err != nil {
+		panic(err)
+	}
+
+	return encoded
+}
+
+func (e *Estimator) Estimate(ctx context.Context, provider *ethrpc.Provider, walletConfig WalletConfig, walletContext WalletContext, txs Transactions) (uint64, error) {
+	address, err := AddressFromWalletConfig(walletConfig, walletContext)
+	if err != nil {
+		return 0, err
+	}
+
+	isEOA, err := e.AreEOAs(ctx, provider, walletConfig)
+	if err != nil {
+		return 0, err
+	}
+
+	willSign, err := e.PickSigners(ctx, walletConfig, isEOA)
+	if err != nil {
+		return 0, err
+	}
+
+	signature := e.BuildStubSignature(walletConfig, willSign, isEOA)
+
+	overwrites := map[common.Address]*Overwrite{
+		walletContext.MainModuleAddress: {
+			Code: contracts.WalletGasEstimatoreDeployedBytecode,
+		},
+		walletContext.MainModuleUpgradableAddress: {
+			Code: contracts.WalletGasEstimatoreDeployedBytecode,
+		},
+	}
+
+	estimates := make([]*big.Int, len(txs)+1)
+
+	// Compute gas estimation for slices of all transactions
+	// including no transaction execution and all transactions
+
+	nonce, err := txs.Nonce()
+	if err != nil {
+		return 0, err
+	}
+
+	for i := range estimates {
```

Comment:
Ideally we could also do this in parallel via multicall, but that's a further optimization

Reply:
Same as before, once we have a ton of tests we can play around with those optimisations